icepyx/core/convert.py

```diff
@@ -0,0 +1,209 @@
+        # from h5py to zarr: use zarr default compression options
+        kws.setdefault('fill_value', source.fillvalue)
+        # create new dataset in destination
+        ds=dest.create_dataset(name,shape=source.shape,dtype=source.dtype,**kws)
+        # copy data going chunk by chunk to avoid loading in entirety
+        shape = ds.shape
+        chunks = ds.chunks
+        chunk_offsets = [range(0, s, c) for s, c in zip(shape, chunks)]
+        for offset in itertools.product(*chunk_offsets):
+            sel = tuple(slice(o, min(s, o + c)) for o, s, c in
+                zip(offset, shape, chunks))
+            ds[sel] = source[sel]
+        # copy attributes
+        attrs = {key:attributes_encoder(source.attrs[key]) for key in
+            source.attrs.keys() if attributes_encoder(source.attrs[key])}
+        ds.attrs.update(attrs)
+    else:
+        # copy a group
+        if (dest is not None and name in dest and hasattr(dest[name], 'shape')):
+            raise CopyError('an array {!r} already exists in destination '
+                '{!r}'.format(name, dest.name))
+        # require group in destination
+        grp = dest.require_group(name)
+        # copy attributes
+        attrs = {key:attributes_encoder(source.attrs[key]) for key in
+            source.attrs.keys() if attributes_encoder(source.attrs[key])}
+        grp.attrs.update(attrs)
+        # recursively copy from source
+        for k in source.keys():
+            copy_from_HDF5(source[k], grp, name=k)
+
+# PURPOSE: reduce HDF5 files to beam groups and output to ascii
+def HDF5_to_ascii(hdf5_file, path, reformat):
+    """
+    convert the HDF5 file to beam-level ascii files copying reduced sets of data
+    """
+    # split extension from HDF5 file
+    fileBasename,fileExtension = os.path.splitext(hdf5_file.filename)
+    delimiter = ',' if reformat == 'csv' else '\t'
+    # copy bare minimum variables from the HDF5 file to the zarr file
+    source = h5py.File(hdf5_file,mode='r')
+    # compile regular expression operator for extracting info from ICESat2 files
+    rx = re.compile(r'(processed)?(ATL\d+)(-\d{{2}})?_(\d{4})(\d{2})(\d{2})'
+        r'(\d{2})(\d{2})(\d{2})_(\d{4})(\d{2})(\d{2})_(\d{3})_(\d{2})(.*?).h5$')
+    # extract parameters from ICESat2 HDF5 file
+    SUB,PRD,HEM,YY,MM,DD,HH,MN,SS,TRK,CYCL,GRAN,RL,VERS,AUX = \
+        rx.findall(os.path.basename(hdf5_file.filename)).pop()
+    # find valid beam groups by testing for particular variables
+    if (PRD == 'ATL06'):
+        VARIABLE_PATH = ['land_ice_segments','segment_id']
+    elif (PRD == 'ATL07'):
+        VARIABLE_PATH = ['sea_ice_segments','height_segment_id']
+    elif (PRD == 'ATL08'):
+        VARIABLE_PATH = ['land_segments','segment_id_beg']
+    elif (PRD == 'ATL10'):
+        VARIABLE_PATH = ['freeboard_beam_segments','delta_time']
+    elif (PRD == 'ATL12'):
+        VARIABLE_PATH = ['ssh_segments']['delta_time']
+    # create list of valid beams within the HDF5 file
+    beams = []
```

Comment:
Can we simplify this by using some of the existing variable module functionality? That might also make it easier to generalize.